### PR TITLE
fix(payment): INT-4481 GooglePay: Fix issue when attempting to pay without a last name provided in the billing address

### PR DESCRIPTION
**Pull request by roberbnd**
## What?  [INT-4481](https://jira.bigcommerce.com/browse/INT-4481)
Changes the way to get the first name

## Why?
When the credit/debit card hasn't the last name the error was happening
image

## Testing / Proof
**Before:** https://drive.google.com/file/d/1seY1J_tMoYWe6PA7AQYjZxiUu7bg44GA/view?usp=sharing
**After:** https://drive.google.com/file/d/1jiB8BW8eUM_20wDmuQ8KAPCGyw4ADMKX/view?usp=sharing

@bigcommerce/checkout @bigcommerce/payments @bigcommerce/apex-integrations

### CODE REVIEW

**Comment by roberbnd** [comment]
@vmparra @clopezh Could you review the code? please

**Review by vmparra — COMMENTED** [review]
LGTM 👍🏻

File: src/payment/strategies/googlepay/googlepay-payment-processor.spec.ts [hunk 1]
```diff
@@ -238,6 +238,21 @@ describe('GooglePayPaymentProcessor', () => {
             expect(googlePayInitializer.parseResponse).toHaveBeenCalled();
         });
 
+        it('handles success response without last name', async () => {
+            jest.spyOn(googlePayInitializer, 'parseResponse').mockReturnValue(tokenizePayload);
+            jest.spyOn(requestSender, 'post').mockReturnValue(Promise.resolve());
+            jest.spyOn(store.getState().billingAddress, 'getBillingAddress').mockReturnValue(getGooglePayAddressMock());
+            jest.spyOn(billingAddressActionCreator, 'updateAddress');
```
Comment (mauricio-sg) [inline src/payment/strategies/googlepay/googlepay-payment-processor.spec.ts:245]: No need to spy on this.

**Review by mauricio-sg — APPROVED** [review]

**Review by clopezh — APPROVED** [review]
LGTM 👍🏻

File: src/payment/strategies/googlepay/googlepay-payment-processor.ts [hunk 2]
```diff
@@ -156,11 +156,12 @@ export default class GooglePayPaymentProcessor {
     }
 
     private _mapGooglePayAddressToBillingAddress(paymentData: GooglePaymentData, id: string): BillingAddressUpdateRequestBody {
-        const firstName = paymentData.paymentMethodData.info.billingAddress.name.split(' ').slice(0, -1).join(' ');
+        const name = paymentData.paymentMethodData.info.billingAddress.name.split(' ');
```
Comment (leeBigCommerce) [inline src/payment/strategies/googlepay/googlepay-payment-processor.ts:159]: People won't expect an array if we use a singular noun.

It's also a bit odd to split the names here.

Given the array isn't used anywhere else in this scope, it should probably happen within `_getFirstAndLastName`.

I'd recommend losing the split (putting that in the helper function) and moving the use of `_getFirstAndLastName` right next to it:

```
        const fullName = paymentData.paymentMethodData.info.billingAddress.name;
        const [firstName, lastName] = this._getFirstAndLastName(fullName);
```

File: src/payment/strategies/googlepay/googlepay-payment-processor.ts [hunk 3]
```diff
@@ -232,4 +233,12 @@ export default class GooglePayPaymentProcessor {
             this._billingAddressActionCreator.updateAddress(googlePayAddressMapped)
         );
     }
+
+    private _getFirstAndLastName(name: string[]): string[] {
```
Comment (leeBigCommerce) [inline src/payment/strategies/googlepay/googlepay-payment-processor.ts:237]: When defining a tuple return type it's best to explicitly declare it like:

`[string, string]`

That way people see an error if they try to deconstruct elements that don't exist:

`const [firstName, lastName, somethingElse] = this._getFirstAndLastName(name);`

File: src/payment/strategies/googlepay/googlepay-payment-processor.ts [hunk 4]
```diff
@@ -232,4 +233,12 @@ export default class GooglePayPaymentProcessor {
...
+        if (name.length > 1) {
+            return [name.slice(0, -1).join(' '), ...name.slice(-1)];
+        }
+
+        return [name.join(' '), ''];
```
Comment (leeBigCommerce) [inline src/payment/strategies/googlepay/googlepay-payment-processor.ts:242]: In this scenario we're rejoining a string we've only just split, by starting with the whole name still as a string we can simply reuse it:

```
    private _getFirstAndLastName(fullName: string): [string, string] {
        const nameParts = fullName.split(' ');

        if (nameParts.length === 1) {
            return [fullName, ''];
        }

        const firstName = nameParts.slice(0, -1).join(' ');
        const lastName = nameParts[nameParts.length - 1];

        return [firstName, lastName];
    }
```

Comment (leeBigCommerce) [reply src/payment/strategies/googlepay/googlepay-payment-processor.ts:242]: This function also deserves its own tests, given that is covers a few scenarios

```
_getFirstAndLastName('Lee Dunkley');
_getFirstAndLastName('Cher');
_getFirstAndLastName('Neil deGrasse Tyson');
```

Comment (leeBigCommerce) [reply src/payment/strategies/googlepay/googlepay-payment-processor.ts:242]: Why not also use it in `_mapGooglePayAddressToShippingAddress` ?

File: src/payment/strategies/googlepay/googlepay-payment-processor.spec.ts [hunk 5]
```diff
@@ -238,6 +238,20 @@ describe('GooglePayPaymentProcessor', () => {
             expect(googlePayInitializer.parseResponse).toHaveBeenCalled();
         });
 
+        it('handles success response without last name', async () => {
+            jest.spyOn(googlePayInitializer, 'parseResponse').mockReturnValue(tokenizePayload);
+            jest.spyOn(requestSender, 'post').mockReturnValue(Promise.resolve());
+            jest.spyOn(store.getState().billingAddress, 'getBillingAddress').mockReturnValue(getGooglePayAddressMock());
+            const googlePaymentDataMock = getGooglePaymentDataMock();
+            googlePaymentDataMock.paymentMethodData.info.billingAddress = getGooglePayAddressMock();
```
Comment (leeBigCommerce) [inline src/payment/strategies/googlepay/googlepay-payment-processor.spec.ts:246]: Why are we patching a mock?

Could `getGooglePaymentDataMock` not contain a billing address by default?

If not, maybe it's worth having two different mock generating function with more specific names

**Comment by roberbnd** [comment]
@leeBigCommerce could you review?

File: src/payment/strategies/googlepay/googlepay-payment-processor.spec.ts [hunk 6]
```diff
@@ -238,6 +238,56 @@ describe('GooglePayPaymentProcessor', () => {
...
+
+            expect(googlePayInitializer.parseResponse).toHaveBeenCalled();
+        });
+
+        it('handles success response with 3 word in fullname', async () => {
+            jest.spyOn(googlePayInitializer, 'parseResponse').mockReturnValue(tokenizePayload);
+            jest.spyOn(requestSender, 'post').mockReturnValue(Promise.resolve());
+            jest.spyOn(store.getState().billingAddress, 'getBillingAddress').mockReturnValue(getGooglePayAddressMock());
+            const googlePaymentDataMock = getGooglePaymentDataMock();
+            googlePaymentDataMock.paymentMethodData.info.billingAddress.name = 'mock mock mock';
```
Comment (leeBigCommerce) [inline src/payment/strategies/googlepay/googlepay-payment-processor.spec.ts:271]: These tests don't prove that `_getFirstAndLastName` works as intended. If we change this value from `mock mock mock` to something else, the test still passes. 

Instead, if we make `getFirstAndLastName` an imported utility dependancy of this file, we can test it more directly.

E.g. something like:

```
import { getFirstAndLastName } from './get-first-and-last-name';

describe('With a one part name', () => {
    it('assigns the whole string as the first name', () => {
        expect(getFirstAndLastName('Cher')).toEqual(['Cher', '']);
    });
});

describe('With a two part name', () => {
    it('splits the string into a first and a second name', () => {
        expect(getFirstAndLastName('Lee Dunkley')).toEqual(['Lee', 'Dunkley']);
    });
});

describe('With a three or more part name', () => {
    it('assigns the final part as the last name and all others as the first name', () => {
        expect(getFirstAndLastName('Neil deGrasse Tyson')).toEqual(['Neil deGrasse', 'Tyson']);
        expect(getFirstAndLastName('Ignacio Arsenio Travieso Scull')).toEqual(['Ignacio Arsenio Travieso', 'Scull']);
    });
});
```

Then this file (the class' tests) can be ignorant of all the permutations and just test one representative variant.

**Comment by roberbnd** [comment]
@leeBigCommerce Are you agree? ♻️

**Review by leeBigCommerce — APPROVED** [review]
In future please use commit comments to describe what is being achieved via the code changes, not who suggest that you make them.

For now, just be sure to squash and merge 👍

File: src/payment/strategies/googlepay/googlepay-get-first-and-last-name.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+export function getFirstAndLastName(fullName: string): [string, string] {
+    const nameParts = fullName.split(' ');
+
+    if (nameParts.length === 1) {
+        return [fullName, ''];
+    }
+
+    const firstName = nameParts.slice(0, -1).join(' ');
```
Comment (animesh1987) [inline src/payment/strategies/googlepay/googlepay-get-first-and-last-name.ts:8]: Why are we doing this?

**Review by animesh1987 — APPROVED** [review]

**Comment by JuanTorres077** [comment]
💚 Tested
QA Details: Below providers were tested  along with Google Pay : 
Adyen 
Stripe 
Authorize.Net 
Paypal 
Checkout. com 
For further details please take a look:
[1174.pdf](https://github.com/bigcommerce/checkout-sdk-js/files/6896672/1174.pdf)

**Review by vmparra — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**